include/deal.II/grid/grid_tools.h

```diff
@@ -2142,6 +2162,453 @@ namespace GridTools
+        //
+        // Given \xi_k, find \delta\xi_k so that
+        //   H_k \delta\xi_k = - F_k
+        // where H_k is an approximation to the second derivatives of J at \xi_k,
+        // and F_k is the first derivative of J.  We'll iterate this a number of
+        // times until the right hand side is small enough. As a stopping
+        // criterion, we terminate if ||\delta\xi||<eps.
+        //
+        // As for the Hessian, the best choice would be
+        //   H_k = J''(\xi_k)
+        // but we'll opt for the simpler Gauss-Newton form
+        //   H_k = A^T A
+        // i.e.
+        //   (H_k)_{nm} = \sum_{i,j} v_i*v_j *
+        //                   \partial_n phi_i *
+        //                   \partial_m phi_j
+        // we start at xi=(0.5, 0.5).
+        Point<structdim> xi;
+        for (unsigned int d=0; d<structdim; ++d)
+          xi[d] = 0.5;
+
+        Point<spacedim> x_k;
+        for (unsigned int i=0; i<GeometryInfo<structdim>::vertices_per_cell; ++i)
+          x_k += object->vertex(i) *
+                 GeometryInfo<structdim>::d_linear_shape_function (xi, i);
+
+        do
+          {
+            Tensor<1,structdim> F_k;
+            for (unsigned int i=0; i<GeometryInfo<structdim>::vertices_per_cell; ++i)
+              F_k += (x_k-trial_point)*object->vertex(i) *
+                     GeometryInfo<structdim>::d_linear_shape_function_gradient (xi, i);
+
+            Tensor<2,structdim> H_k;
+            for (unsigned int i=0; i<GeometryInfo<structdim>::vertices_per_cell; ++i)
+              for (unsigned int j=0; j<GeometryInfo<structdim>::vertices_per_cell; ++j)
+                {
+                  Tensor<2, structdim> tmp = outer_product(
+                                               GeometryInfo<structdim>::d_linear_shape_function_gradient(xi, i),
+                                               GeometryInfo<structdim>::d_linear_shape_function_gradient(xi, j));
+                  H_k += (object->vertex(i) * object->vertex(j)) * tmp;
+                }
+
+            const Tensor<1,structdim> delta_xi = - invert(H_k) * F_k;
+            xi += delta_xi;
+
+            x_k = Point<spacedim>();
+            for (unsigned int i=0; i<GeometryInfo<structdim>::vertices_per_cell; ++i)
+              x_k += object->vertex(i) *
+                     GeometryInfo<structdim>::d_linear_shape_function (xi, i);
+
+            if (delta_xi.norm() < 1e-7)
+              break;
+          }
+        while (true);
+
+        return x_k;
+      }
+    }
+  }
+
+
+
+  template <typename Iterator>
+  Point<Iterator::AccessorType::space_dimension>
+  project_to_object(const Iterator &object,
+                    const Point<Iterator::AccessorType::space_dimension> &trial_point)
+  {
+    const int spacedim = Iterator::AccessorType::space_dimension;
+    const int structdim = Iterator::AccessorType::structure_dimension;
+
+    Point<spacedim> projected_point = trial_point;
+
+    if (spacedim <= structdim)
```

Comment:
I tend to put the fixed quantities on the right hand side of a comparison and the variables to the left -- so would have written this as `structdim >= spacedim`. But this would have also made it clearer that the `>` can never happen. So maybe write this as `structdim == spacedim`?